### DISCUSSION TITLE
NTP Synchronization Parameters in UnetStack

### DISCUSSION THREAD

**Discussion by brodiealexander** [org-arl/unet-support#3] [Q&A] [answered]
Howdy! My team has a use case where we have an NTP server running on our private intranet. A few blog posts mention that the RTC is synchronized using NTP, but does not talk about any mechanism to specify time servers. Is this a supported use case? If not, is this something you might be willing to support in a future version of UnetStack?

Currently, we synchronize time using a Python script which sets phy.rtc. However, this method is susceptible to latency and jitter when working over WiFi. 

Thanks,
     Brodie

**Comment by notthetup** [comment, marked as answer]
Hey @brodiealexander! 

UnetStack uses a couple of clocks. Most general time keeping is done using the system clock. This is the clock of the underlying OS. For anything to do with the comms system, it keeps a monotonic clock called the "Physical Layer Clock" which it derives from the data acquisition system. All the comms message timestamps are based on this "Physical Layer Clock". 

For UnetStack running on @subnero1 modems, you can set the system time in a couple of ways..

- Using NTP. The modems run an NTP client, which can connect to global NTP servers to get time. 
- Using the WebUI. Click on the "sync time" button
- Using an API. Set the `device.time` (Gen 4 gardware) or `phy.rtc` (Gen 3 hardware) parameter as you're doing.

For the NTP client, we default to the standard "pool.ntp.org" time servers. Devices with UnetStack 4 supports viewing and adding to this pool using the `ntp` and `ntpadd` shell extensions.

```
### ntpadd - add NTP server

Usage:
  ntpadd [ip, [port]]

Examples:
  ntpadd '192.168.1.214'                  // add NTP server
  ntpadd '192.168.1.214', 123             // add NTP server with port 123

### ntp - list NTP servers

Examples:
  ntp                                     // list all NTP servers
```

**Reply by brodiealexander** [reply]
Thanks @notthetup !
